### PR TITLE
class-generator: fix when kind do not have `properties`

### PR DESCRIPTION
**Pull request by myakove**
1. Do not fail when the resource does not have `properties` in the schema.
2. Do not overwrite existing files when the schema is generated with older cluster versions.

## Summary by CodeRabbit

- **New Features**
	- Enhanced version management for schema generation, including tracking of the last generated cluster version.
	- Expanded list of supported Kubernetes and OpenShift resource types, including new entries for various components and services.
	- Introduced new JSON schema definitions for multiple resources within the Tekton and OpenShift Pipelines frameworks.
	- New configuration file `.coderabbit.yaml` to manage project settings, review processes, and tool integrations.

- **Bug Fixes**
	- Improved handling of schema properties to prevent errors when expected structures are missing.

- **Chores**
	- Updated `.gitignore` to include new entries for ignored files and directories, streamlining version control.
	- Added a new version identifier for the cluster component.

### CODE REVIEW

**Comment by redhat-qe-bot1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
    * You can add extra args to the Podman build command
        * Example: `/build-and-push-container --build-arg OPENSHIFT_PYTHON_WRAPPER_COMMIT=<commit_hash>`
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest python-module-install`: Retest python-module-install
 * `/retest all`: Retest all

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: class_generator/class_generator.py [hunk 1]
```diff
@@ -177,14 +179,39 @@ def update_kind_schema():
...
+    cluter_version_file = Path("class_generator/__cluster_version__.txt")
+
+    with open(cluter_version_file, "r") as fd:
+        last_cluster_version_generated = fd.read().strip()
+
+    newer_version: bool = Version(cluster_version) > Version(last_cluster_version_generated)
+
+    if newer_version:
+        with open(cluter_version_file, "w") as fd:
+            fd.write(cluster_version)
```
[the remark this answers is not in the document]
Comment (myakove) [reply class_generator/class_generator.py:209]: done

File: class_generator/class_generator.py [hunk 2]
```diff
@@ -177,14 +179,39 @@ def update_kind_schema():
...
+    if newer_version:
+        # copy all files from tmp_schema_dir to schema dir
+        shutil.copytree(src=tmp_schema_dir, dst=SCHEMA_DIR, dirs_exist_ok=True)
+
+    else:
+        # Copy only new files from tmp_schema_dir to schema dir
+        for root, _, files in os.walk(tmp_schema_dir):
+            for file_ in files:
+                if not os.path.isfile(Path(SCHEMA_DIR) / file_):
+                    shutil.copy(src=Path(root) / file_, dst=Path(SCHEMA_DIR) / file_)
```
[the remark this answers is not in the document]
Comment (myakove) [reply class_generator/class_generator.py:229]: tmp_schema_dir dir contains only json files, no directories

File: class_generator/class_generator.py [hunk 3]
```diff
@@ -177,14 +179,39 @@ def update_kind_schema():
...
-    ocp_openapi_json_file = f"class_generator/__k8s-openapi-{cluster_version}__.json"
+    ocp_openapi_json_file = f"/tmp/__k8s-openapi-{cluster_version}__.json"
+    last_cluster_version_generated: str = ""
+    cluter_version_file = Path("class_generator/__cluster_version__.txt")
+
+    with open(cluter_version_file, "r") as fd:
+        last_cluster_version_generated = fd.read().strip()
+
+    newer_version: bool = Version(cluster_version) > Version(last_cluster_version_generated)
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply class_generator/class_generator.py:206]: This file is part of the repository and always exists, we want to fail if not

File: class_generator/schema/manualapprovalgatelist.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      "description": "APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources",
+      "type": "string"
+    },
+    "items": {
+      "description": "List of manualapprovalgates. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md",
+      "type": "array",
+      "items": {
+        "$ref": "_definitions.json#/definitions/dev.tekton.operator.v1alpha1.ManualApprovalGate"
+      }
+    },
```
[the remark this answers is not in the document]
Comment (myakove) [reply class_generator/schema/manualapprovalgatelist.json:18]: all files under `class_generator/schema/` are auto-generated

**Comment by myakove** [comment]
@coderabbitai configuration

File: class_generator/class_generator.py [hunk 5]
```diff
@@ -73,7 +71,7 @@ def _is_kind_and_namespaced(client: str, _key: str, _data: Dict[str, Any]) -> Di
     return not_resource_dict
 
 
-def map_kind_to_namespaced(client: str):
+def map_kind_to_namespaced(client: str, newer_cluster_version: bool, schema_definition_file):
```
Comment (rnetser) [inline class_generator/class_generator.py:74]: `schema_definition_file` missing type and return type

**Review by rnetser — CHANGES_REQUESTED** [review]

File: class_generator/class_generator.py (hunk 5 above)
Comment (myakove) [reply class_generator/class_generator.py]: Done

**Review by rnetser — APPROVED** [review]

File: class_generator/class_generator.py [hunk 6]
```diff
@@ -177,15 +193,42 @@ def update_kind_schema():
...
+
+    with open(cluster_version_file, "r") as fd:
+        last_cluster_version_generated = fd.read().strip()
+
+    newer_version: bool = Version(cluster_version) > Version(last_cluster_version_generated)
+
+    if newer_version:
+        with open(cluster_version_file, "w") as fd:
+            fd.write(cluster_version)
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply class_generator/class_generator.py:208]: done

File: class_generator/class_generator.py [hunk 7]
```diff
@@ -177,15 +194,42 @@ def update_kind_schema():
 
     cluster_version = get_server_version(client=client)
     cluster_version = cluster_version.split("+")[0]
-    ocp_openapi_json_file = f"class_generator/__k8s-openapi-{cluster_version}__.json"
+    ocp_openapi_json_file = Path(gettempdir()) / f"__k8s-openapi-{cluster_version}__.json"
+    last_cluster_version_generated: str = ""
+    cluster_version_file = Path("class_generator/__cluster_version__.txt")
+
+    with open(cluster_version_file, "r") as fd:
+        last_cluster_version_generated = fd.read().strip()
```
[the remark this answers is not in the document]
Comment (myakove) [reply class_generator/class_generator.py]: Done

**Review by rnetser — APPROVED** [review]

**Review by dbasunag — APPROVED** [review]

**Pull request merged**